FernFlower-Patches/0024-Give-nicer-output-for-float-and-double-literals.patch

```diff
@@ -0,0 +1,244 @@
+ 
+       case CodeConstants.TYPE_DOUBLE:
+         double doubleVal = (Double)value;
+         if (!literal) {
++          // Double constants, some of which can't be represented directly
+           if (Double.isNaN(doubleVal)) {
+             return new FieldExprent("NaN", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+           }
+@@ -203,20 +238,48 @@ public class ConstExprent extends Exprent {
+           else if (doubleVal == Double.MAX_VALUE) {
+             return new FieldExprent("MAX_VALUE", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+           }
++          else if (doubleVal == Double.MIN_NORMAL) {
++            return new FieldExprent("MIN_NORMAL", "java/lang/Float", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
++          }
+           else if (doubleVal == Double.MIN_VALUE) {
+             return new FieldExprent("MIN_VALUE", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+           }
++          // Math constants
++          else if (doubleVal == Math.E) {
++            return new FieldExprent("E", "java/lang/Math", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
++          }
++          else if (PI_DOUBLES.containsKey(doubleVal)) {
++            String[] parts = PI_DOUBLES.get(doubleVal);
++            return getPiDouble(tracer).enclose(parts[0], parts[1]);
++          }
++
++          // Check for cases where a float literal has been upcasted to a double.
++          // (for instance, double d = .01F results in 0.009999999776482582D without this)
++          float nearestFloatVal = (float)doubleVal;
++          if (doubleVal == (double)nearestFloatVal) {
++            // Value can be represented precisely as both a float and a double.
++            // Now check if the string representation as a float is nicer/shorter.
++            // If they're the same, there's no point in the cast and such (e.g. don't decompile 1.0D as (double)1.0F).
++            if (Float.toString(nearestFloatVal).length() < Double.toString(doubleVal).length()) {
++              // Include a cast to prevent using the wrong method call in ambiguous cases.
++              return createFloat(literal, nearestFloatVal, tracer).prepend("(double)");
++            }
++          }
+         }
+-        else if (Double.isNaN(doubleVal)) {
+-          return new TextBuffer("0.0D / 0.0D");
+-        }
+-        else if (doubleVal == Double.POSITIVE_INFINITY) {
+-          return new TextBuffer("1.0D / 0.0D");
+-        }
+-        else if (doubleVal == Double.NEGATIVE_INFINITY) {
+-          return new TextBuffer("-1.0D / 0.0D");
++        else {
++          // Check for special values that can't be used directly in code
++          // (and we can't replace with the constant due to the user requesting not to)
++          if (Double.isNaN(doubleVal)) {
```

Comment:
Clarity of behavior, for the same reason I added the comment (technically not needed, sure, but for a more general set of changes it's something I'd want).  When moving it into a separate method I decided to do that as well because I think it's more readable, but without the separate method it probably isn't entirely needed (if/when I upstream this change I'm probably going to use the separate method version)